AUTOCONFIG/AutoconfigHandler.php

```diff
@@ -0,0 +1,1284 @@
+            'spa'						=> pacol( 1, 		1,      1,      'text', 'Autoconfig_spa'				, ''                                , '' ),
+        );
+    }
+    
+    /**
+     * @param string $username
+     */
+    public function __construct($username) {
+        $this->username = $username;
+        if ( authentication_has_role('admin') ) {
+            $this->is_admin = true;
+            $this->all_domains = list_domains_for_admin( $username );
+            // Get the list of configuration ids, if any, this admin is allowed to access
+            $this->allowed_config_ids = $this->get_config_ids_for_user( $username );
+        }
+    }
+    
+    protected function initMsg() {
+        // Need to develop this part
+    }
+    
+    /**
+     * @return array
+     */
+    public function webformConfig() {
+        return array(
+            # $PALANG labels
+            'formtitle_create' => 'pAutoconfig_page_title',
+            'formtitle_edit' => 'pAutoconfig_page_title',
+            'create_button' => 'save',
+
+            # various settings
+            'required_role' => 'admin',
+            'listview' => 'list-virtual.php',
+            'early_init' => 1, # 0 for create-domain
+        );
+    }
+
+    protected function validate_new_id() {
+        # autoconfig can only be enabled if a domain name exists
+        if ( $this->is_admin ) {
+            if ( count( $this->all_domains ) > 0 ) {
+                return( true );
+            }
+        } else {
+            // Need to develop this part
+            return( true );
+        }
+
+        # still here? This means the mailbox doesn't exist or the admin/user doesn't have permissions to view it
+        $this->errormsg[] = Config::Lang('invalid_parameter');
+        return( false );
+    }
+    
+    public function get_config_ids_for_user($user) {
+        $table_autoconfig = table_by_key('autoconfig');
+        $table_autoconfig_domains = table_by_key('autoconfig_domains');
+        $table_domain_admins = table_by_key('domain_admins');
+        $table_domain = table_by_key('domain');
+        // This is a super admin, so he/she has access to all configs
+        if ( authentication_has_role( 'global-admin' ) ) {
+            // $sql = "SELECT DISTINCT ad.config_id FROM $table_autoconfig_domains ad LEFT JOIN $table_domain d ON ad.domain = d.domain WHERE d.domain != 'ALL AND d.active IS TRUE'";
+            // global admin has access to all config
+            $sql = "SELECT c.config_id FROM $table_autoconfig c";
+        }
+        // This is a per-domain admin, so we use the table domain_admis to cross check which configuration he/she has access
+        elseif ( authentication_has_role( 'admin' ) ) {
+            $E_username = escape_string( $user );
+            $sql = "SELECT DISTINCT ad.config_id FROM $table_domain d LEFT JOIN $table_autoconfig_domains ad ON ad.domain = d.domain WHERE d.active IS TUE AND d.username='$E_username'";
```

Comment:
is this TUE a typo?

Can you rewrite it so there's no chance of $sql being undefined. 
E.g. default $sql to whatever the 'admin' user has ?

Reply:
Good catch. Yes, this should have been `TRUE`

Reply:
( and probably 'table domain_admins' in the comment :+1: